Add u-boot and bootable SD card image creation for Banana Pi R2

dwmw2 opened the pull request:
Add a u-boot build and download the preloader binary from the vendor's github repository, and create a bootable SD card image.

Most of the interesting parts lifted from @lexa2's work.

dwmw2 committed:
mediatek: enable SATA for mt7623

The MT7623 SoC has the same SATA block as the MT7622, so enable it in
MT7623 builds too and add it to the DEVICE_PACKAGES for those boards.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: add mt7623 u-boot build for Banana Pi R2

For building full SD/eMMC images for Banana Pi R2 we'll want a u-boot
image built for that platform.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: new mt7623n preloader package for Banana Pi

Download the boot preloader code from the Banana Pi github repo and make
it available for bootable SD card image creation.

Supports only Banana Pi R2 for now.

Based on work by Alexey Loukianov <lx2@lexa2.ru> and others.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

adrianschmutzler commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -5,6 +5,8 @@ define Device/bpi_bananapi-r2
   DEVICE_DTS := mt7623n-bananapi-bpi-r2
   SUPPORTED_DEVICES := bananapi,bpi-r2
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb-ohci kmod-usb2 kmod-usb3 \
+                     kmod-ata-core kmod-ata-ahci-mtk
```
kmod-usb-core and kmod-ata-core don't need to be specified here, as they will be selected by the other packages anyway.

adrianschmutzler commented on package/boot/mt7623n-preloader/Makefile:
```diff
@@ -0,0 +1,60 @@
+#
+include $(TOPDIR)/rules.mk
+include $(INCLUDE_DIR)/kernel.mk
+#include $(INCLUDE_DIR)/image.mk
```
Commented line can be dropped?

adrianschmutzler commented on target/linux/mediatek/image/gen_banana_pi_img.sh:
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
```
bash needed here, or can this be done with /bin/sh?

dwmw2 replied:
I think it does work with POSIX sh, yes. But then I saw a bunch of other invocations of bash in a quick grep, and didn't bother to change it anyway. Have done now.

adrianschmutzler replied:
Maybe about half a year ago somebody made an effort to reduce the use of bash in our scripts and convert to sh where easily possible. Some fragments are still found in our patchwork, as not all of the patches have been applied.
I'm mainly following that idea.

adrianschmutzler commented on target/linux/mediatek/image/gen_banana_pi_img.sh:
```diff
@@ -0,0 +1,50 @@
+head=4
+sect=63
+
+set `ptgen -o $OUTPUT -h $head -s $sect -l 1024 -t c -p ${BOOTFSSIZE}M -t 83 -p ${ROOTFSSIZE}M -a 0`
```
consider $() instead of backticks

adrianschmutzler commented on target/linux/mediatek/image/gen_banana_pi_img.sh:
```diff
@@ -0,0 +1,50 @@
+dd bs=1024 if="${UBOOT}"     of="${OUTPUT}" seek="${UBOOT_OFFSET}"     conv=notrunc
+dd bs=512  if="${BOOTFS}"    of="${OUTPUT}" seek="${BOOT_OFFSET}"      conv=notrunc
+dd bs=512  if="${ROOTFS}"    of="${OUTPUT}" seek="${ROOTFS_OFFSET}"    conv=notrunc
+
```
Consider removing empty line at EOF ...

adrianschmutzler commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -1,12 +1,34 @@
+		$(STAGING_DIR_IMAGE)/$(UBOOT_TARGET)-u-boot-mtk.bin \
+		$@.boot \
+		$(IMAGE_ROOTFS) \
+                $(CONFIG_BANANA_PI_BOOT_PARTSIZE) \
```
Indent intended? (Even sounds strange when you don't speak it out loud :-) )

dwmw2 committed:
mediatek: add SD card image creation for Banana Pi R2

Based on work by Alexey Loukianov <lx2@lexa2.ru> and others.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 commented:
Thanks for the review @adrianschmutzler. Pushed all those fixes. I'll leave it in [RFC] state for the moment until the replacement SD card arrives today and I can actually test...

adrianschmutzler commented:
> Thanks for the review @adrianschmutzler. Pushed all those fixes. I'll leave it in [RFC] state for the moment until the replacement SD card arrives today and I can actually test...

As you might have realized, I'm limited to drive-by nitpicks here, so yes, maybe keep the RFC until a "proper" reviewer arrives.

cioby23 commented on package/boot/mt7623n-preloader/Makefile:
```diff
@@ -0,0 +1,61 @@
+  FILE:=$(BPI_PRELOADER_PREFIX)-BPI-R2-preloader-2k.img.gz
+  URL:=$(BPI_PRELOADER_URL)
+  URL_FILE:=BPI-R2-preloader-2k.img.gz
+  HASH:=9a961a26be17dd410866a2385fca47b6f0e7cce88eed6248ff5fb145cf935d93
```
I suggest to use this newer preloader https://github.com/BPI-SINOVOIP/BPI-files/blob/master/SD/100MB/**BPI-R2-preloader-DDR1600-20191024-2k.img.gz** instead of the existing one **BPI-R2-preloader-2k.img.gz** since it will allow larger size for u-boot binary (up to 512KB).
Current preloader limits u-boot size to maximum 300KB. I have tested this newer preloader in my OpenWRT fork and it works fine.

dwmw2 replied:
That's excellent; testing now. Thanks.

Now I just need to work out why the kernel isn't receiving its command line from u-boot.

cioby23 replied:
I have observed that u-boot when it's get compiled it generates two binaries one called **u-boot-mtk.bin** and another simply **u-boot.bin**.
When I tried to use the **u-boot-mtk.bin** the kernel does not load and it gets stuck. With the other u-boot.bin binary it works fine.
I compared both of them and I saw the **u-boot-mtk.bin** version when generated injects some extra bytes at the beginning of the file and the preloader does not like it.
I'm not sure if this the same issue you mentioned above.

dwmw2 replied:
I muttered about that one on IRC for a while; it was *one* of the reasons, before the size, that u-boot wasn't working. All fixed now; thanks.

Do you have the command line working?

cioby23 replied:
You refer to the serial console ?

dwmw2 commented:
Building a FIT image instead of a legacy uImage allows U-Boot to set the command line correctly.

dwmw2 commented:
I think I'm happy with this now, although I want to play with making the U-Boot script capable of doing an install to the internal MMC too. @cioby23 you have sysupgrade support in your tree, don't you? Want to look at adding that on top of this series?

dwmw2 committed:
firmware-utils/ptgen: allow explicit placement of partitions

For Banana Pi R2 we need to place the U-Boot partition at precisely 0x50000.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: store MAC address in boot partition on Banana Pi R2

Like many boards, the Banana Pi R2 doesn't have permanant storage of
its MAC address, and we store the first random one that the kernel
generates in order to use it later and at least be consistent.

Store it in the FAT boot partition, just as the U7623 board (and others)
do.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

cioby23 commented:
> I think I'm happy with this now, although I want to play with making the U-Boot script capable of doing an install to the internal MMC too. @cioby23 you have sysupgrade support in your tree, don't you? Want to look at adding that on top of this series?

Yes I have sysupgrade support added in my OpenWRT fork but I did not tested it yet. Maybe sometime next week if I'll have some time I'll try to push it to OpenWRT main repository

dwmw2 commented:
I just changed the partitioning to put u-boot in its own partition so the sysupgrade support may need a small tweak. Also added the MAC address storage in the FAT partition which will also need to be preserved (like the other mt7623 platform does).

I'm working on support for installing to the internal eMMC from u-boot.

lexa2 commented:
@dwmw2 IIRC in my fork sysupgrade was working too but I wasn't really happy with how it works. I was implementing it based on how it is done for other boards with ext4 sysfs support which meant there were two possibilities: full image update using SD image as a sysupgrade file or a root partition update + replacement of kernel file on the boot partition using a special sysupgrade tar.gz image. Neither methods felt good for my taste (despite working fine when tested) as updating on a block-level basis (either full SD image or full root partition) IMO undermines the use of ext4 read-write root filesystem. With this approach user customizations might/would be partly lost between upgrades and that's not good end-user experience. 

The way openwrt sysupgrade is coded best support for conserving custom changes on FS is to use overlayfs, i.e. squashfs+jffs2 instead of ext4 - as that way it is easy to spot and preserve everything, that had been added/changed/removed by end-user on the filesystem.

P.S. After using several BPiR2 boards with OpenWRT I came to conclusion that I'm getting more suitable/manageable results using more sophisticated distros. My current OS of choice for BPiR2 is CentOS 7 AltArch - works perfectly when provided with suitable kernel and is well-known w.r.t. configuration, maintenance and update plan side of things due to being almost identical to CentOS7/RHEL7 installations on full-blown proper servers.

mans0n commented on target/linux/mediatek/image/Makefile:
```diff
@@ -28,6 +28,7 @@ define Device/Default
   IMAGE/sysupgrade.bin := append-kernel | pad-to 128k | append-rootfs | pad-rootfs | append-metadata
   SUPPORTED_DEVICES := $(subst _,$(comma),$(1))
 ifeq ($(SUBTARGET),mt7623)
+  FILESYSTEMS += ext4
```
Could you move this to `Device/bpi_bananapi-r2`? Not all mt7623 devices use ext4.

dwmw2 replied:
Not quite sure why the Unielec board can't use ext4; it seems almost identical in many respects.

mans0n replied:
Probably Unielec can use ext4, but its image has not been tested with sysupgrade, right? And new nand-based device might be added in the future, which does not rely on ext4.

dwmw2 replied:
I've ordered a Unielec board so I can test it. It looks like a lot of unification can be done — including using a FIT image instead of a legacy uImage.

I'm less bothered about a hypothetical possibility that there will be future boards which don't have block devices. None of the current setup would work for those anyway.

But I don't care that much about ext4. @lexa2 is right that the upgrade process is a bit suboptimal with a full read/write file system. You still need to explicitly mark the files you want preserved over upgrade in `/lib/keep.d` anyway and you're probably better off with squashfs + f2fs overlay.

mans0n replied:
I've actually made a support patch for a nand-based device and some of it is in my PR. #3066 But you're right, you don't really need to care about the nand stuff here.

If you can test this on the Unielec board then it would be just fine. I'm not familiar with f2fs so I'll just leave it to you.

mans0n commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -1,11 +1,39 @@
+  DEVICE_PACKAGES := kmod-usb-ohci kmod-usb2 kmod-usb3 kmod-ata-ahci-mtk \
+                     mt7623n_preloader u-boot-mt7623n_bpir2 mkf2fs
+  UBOOT_TARGET := mt7623n_bpir2
+  IMAGES += sdcard.bin.gz
```
Does it still use sysupgrade.bin? Does the sysupgrade image still boot well?

dwmw2 replied:
We're just working on forward porting the sysupgrade support.

dwmw2 replied:
No, I've removed sysupgrade.bin and we upgrade using the same SD image.

mans0n commented on target/linux/mediatek/image/gen_banana_pi_img.sh:
```diff
@@ -0,0 +1,148 @@
+\x00\x08\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+
+case $1 in
+    sd)
```
It's a matter of personal preference but I'd rather remove the indent as most [02_network](https://github.com/openwrt/openwrt/blob/master/target/linux/mpc85xx/base-files/etc/board.d/02_network#L13) do. And I would be happy if you unify all indents in this file into space or tab...

mans0n commented on target/linux/mediatek/image/gen_banana_pi_img.sh:
```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env sh
```
It seems there's no usage of `#!/usr/bin/env sh` in the repo except for `tools/automake/files/aclocal`. I guess `#!/bin/sh` is enough here...

mans0n commented on target/linux/mediatek/Makefile:
```diff
@@ -6,13 +6,14 @@ ARCH:=arm
 DEFAULT_PACKAGES += \
-	kmod-leds-gpio kmod-gpio-button-hotplug
+	kmod-leds-gpio kmod-gpio-button-hotplug \
+	uboot-envtools partx-utils e2fs-progs
```
Do you mean "e2fsprogs"?

mans0n commented on target/linux/mediatek/mt7623/base-files/lib/preinit/07_set_iface_mac:
```diff
@@ -29,17 +21,37 @@ preinit_set_mac_address() {
+preinit_set_mac_address() {
 
-		mac=$(macaddr_add $mac 1)
+	. /lib/functions.sh
```
Could you move the include lines to outside of function? [ref](https://github.com/openwrt/openwrt/blob/master/target/linux/mvebu/base-files/lib/preinit/79_move_config#L4)

mans0n commented on target/linux/mediatek/mt7623/base-files/lib/preinit/07_set_iface_mac:
```diff
@@ -29,17 +21,37 @@ preinit_set_mac_address() {
+	unielec,u7623-02-emmc-512m)
+		set_recovery_mac_address /dev/mmcblk0p1
+		;;
+	bananapi,bpi-r2)
```
`bananapi,bpi-r2` should go above `unielec,u7623-02-emmc-512m`.

mans0n commented:
I'm still not convinced about adding ext4 feature to entire mediatek target...

dwmw2 committed:
mediatek/mt7623: unify features and packages, add ext4 and usb

The supported MT7623 boards are mostly identical (what with being a
System-on-Chip and all), so unify the DEVICE_PACKAGES for them, and add
ext4 and usb support for them.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: Implement sysupgrade support for Banana Pi R2

Based on work by Alexey Loukianov <lx2@lexa2.ru> and others.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 commented:
Added sysupgrade support. Haven't tested the mt7632-emmc support; there is a board on the way but it'll be a while.

dwmw2 commented:
@kristrev please could I ask you to test that I haven't broken the Unielec U7632 upgrade?

cioby23 commented on target/linux/mediatek/mt7623/base-files/lib/preinit/07_set_iface_mac:
```diff
@@ -1,24 +1,20 @@
 #!/bin/sh
```
The shebang line must be removed according to this commit https://github.com/openwrt/openwrt/commit/48c1fdd046f409fd5e28932f1d53343fa214fbc4

cioby23 commented on target/linux/mediatek/mt7623/base-files/lib/preinit/07_set_iface_mac:
```diff
@@ -1,24 +1,20 @@
 #!/bin/sh
```
The shebang line must be removed according to this commit https://github.com/openwrt/openwrt/commit/48c1fdd046f409fd5e28932f1d53343fa214fbc4

dwmw2 replied:
The shebang line *was* removed by commit 48c1fdd. This isn't a new file in my PR.

cristian-ciobanu reviewed (commented):
The mt7623 kernel config was recently updated and changes were pushed in this commit https://github.com/openwrt/openwrt/commit/c4037a5b9eee09c96e5c8108f1b7aa439f96cb6b which conflicts with the changes from made in your pull request.

dwmw2 commented:
> The mt7623 kernel config was recently updated and changes were pushed in this commit [c4037a5](https://github.com/openwrt/openwrt/commit/c4037a5b9eee09c96e5c8108f1b7aa439f96cb6b) which conflicts with the changes from made in your pull request.

Dropped that patch.

mans0n commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -1,5 +1,9 @@
 KERNEL_LOADADDR := 0x80008000
 
+FEATURES += ext4 usb
```
Are you sure? Usually FEATURES is in target/subtarget Makefile, not image Makefile.
And AFAIK DEVICE_PACKAGES is a device-specific variable so it should be defined in device recipes.

adrianschmutzler replied:
DEVICE_PACKAGES is device-specific, but defining it like below still works if none of the devices has it defined (which seems to be the case here).

adrianschmutzler commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -1,5 +1,9 @@
 
+FEATURES += ext4 usb
+DEVICE_PACKAGES := kmod-fs-vfat kmod-nls-cp437 kmod-nls-iso8859-1 kmod-mmc \
+	mkf2fs e2fsprogs kmod-usb-ohci kmod-usb2 kmod-usb3
```
Ah, GitHub messed up the patch order again.

Anyway, DEVICE_PACKAGES should be defined per device. It won't work properly otherwise.
If you want to provide default packages for one subtarget, you have different options:
1. use the Device/Default definition in image/Makefile, which will be used as base setup for all devices (there already is an ifeq for a subtarget there)
2. create a new definition that will be used for mt7623 devices only and then included for the devices * (however, that's not really helpful just for one variable)
3. use DEFAULT_PACKAGES in mediatek/Makefile or mt7623/target.mk
3. Just keep the DEVICE_PACKAGES with the devices. Particularly for usb, I'd personally go that way.

dwmw2 replied:
> 2.  create a new definition that will be used for mt7623 devices only and then included for the devices (see below, however, that's not really helpful just for one variable)

But then the other variables from the ifeq($(SUBTARGET),mt7623) block in `image/Makefile` could live there too, right? And the same variables in the `else` block could go into the other `mt76xx.mk` files close to the actual devices?

adrianschmutzler replied:
One could go that road, yes, however note that you would need to have the "include" $(Device/mtsomething) for every devices. With the ifeq in in Device/Default, this is done automagically ...

I'd address from a conceptual point of view: Is it really a "default" for all/most devices? Then you should go with option 1 and/or 3. Is it a pattern that will be relevant for a class/group of devices, but not the entire subtarget (as far as you can predict)? Then go with 2. Is it actually device-specific, but happens to be the same for all relevant devices by coincidence? Then I'd tend to use option 4.
Of course, one may combine these options e.g. for different variables or packages ...

I'm aware that this is quite hard to decide if there is only two devices in the subtarget ...

Personally, I'm not a big fan of option 2, as this seems nice at the beginning, but will bite back at end (mostly because you add another layer between default device and device definition).

dwmw2 replied:
Yeah, screw it. Someone is already adding a NAND-based mt7623 device anyway. I'll live with the duplication in `DEVICE_PACKAGES` for now. Thanks.

cioby23 commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -1,12 +1,41 @@
+  IMAGE/img.gz += banana-pi-sdcard | gzip | append-metadata
+  KERNEL := kernel-bin | fit none $$(DTS_DIR)/$$(DEVICE_DTS).dtb
+  KERNEL_INITRAMFS :=
 endef
```
Right now I see only **squashfs** images are generated. Can this line **FILESYSTEMS += ext4** be added in the device definition section to be able to generate ext4 images for this target also ?

dwmw2 replied:
I thought there was pushback against that. So I dropped it and we can add it later perhaps after we've done the basic support. And once the sysupgrade support is tested.

cioby23 replied:
I have cloned the OpenWRT trunk, applied all your changes from this pull request and compiled a new ext4 image. I have wrote the image on the SD card but when trying to browse on the Internet every page loads very slow. It takes more than 2 minutes and some of them time out. Ping works fine (I left all settings at the defaults)
I saw this problem appeared a while ago after kernel 5.4.32 was released and something has been broken since then because on my OpenWRT fork using kernel 5.4.31 everything runs fine. If you can test please let me know if you experience the same issue.

adrianschmutzler replied:
Hmm, interesting. That FILESYSTEMS is not listed in DEFAULT_DEVICE_VARS. Either it's hidden somewhere or defining it per-device will create problems ...

dwmw2 replied:
Most of my testing has been the wan device, copying an image to the device for sysupgrade. It hasn't seemed slow. Can you do a packet capture and see what's happening? Are there dropped packets, bad checksums?

https://github.com/lexa2/openwrt/commit/5ec02dbe03958e39db292ca4ac7d94bfab9aea72 comes to mind.

dwmw2 replied:
Also, are you seeing this with IPv6 or Legacy IP or both?

cioby23 replied:
Yes with both protocols IPv4 or IPv6 is the same. I can barely get 5 Mbps in both directions. I have a 500Mbps/250Mbps connection. I did not apply that patch lexa2@5ec02db in my fork it works. No dropped packets or any other interface errors on **wan** or **lan** interfaces only the speed is throttled.

dwmw2 replied:
What is the network topology? Is this routing *through* the Banana Pi? Just routing, or also NAT.

Either way I'd be surprised if this PR makes the difference; can you test the OpenWrt build (with its convoluted install instructions) from current master?

cioby23 replied:
Right now my Banana-Pi R2 router is behind another TP-Link router which has the public IP on it. The WAN port (DHCP) is connected into the TP-link router and the LAN ports are bridged. I'm using masquerading/NAT to access Internet.
I compiled the image yesterday from OpenWRT master and applied the commits from your pull request and the situation it's the same.

dwmw2 replied:
I've plugged a client into the a LAN port of mine (which was still configured for NAT as it's a clean installation) and run iperf to a machine on the house network, so it's mostly like your setup. I'm currently connected to an audio conference call through it, which is working perfectly and iperf reports 660Mb/s TCP at the same time.
```
[  3] local 192.168.1.228 port 36468 connected with 90.155.92.213 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-30.0 sec  2.31 GBytes   660 Mbits/sec
```
>  I compiled the image yesterday from OpenWRT master and applied the commits from your pull request and the situation it's the same.

The idea was to test it *without* my commits, so that we know it isn't introduced by them and we can have this conversation elsewhere and actually merge this PR :)

dwmw2 replied:
> Update: the problem appears to be on the LAN side. I ran wget directly on the box and download several iso images and the speed is fine only when trying from the LAN things go worse.

Do you have multiple devices in LAN ports? Can you unplug all but one and does it make a difference? Not that it should, but that's the main difference I can think of between your test and mine...

cioby23 replied:
No I only have my desktop connected behind it. I also did an iperf test and I got around 900 Mbps from my PC to BananaPi R2 but in the opposite direction I barely get few kilobytes
root@OpenWrt:/# iperf3 -c 192.168.1.220
Connecting to host 192.168.1.220, port 5201
[  5] local 192.168.1.1 port 60766 connected to 192.168.1.220 port 5201
[ ID] Interval                 Transfer        Bitrate               Retr
[  5]   0.00-10.00  sec  45.6 KBytes  37.4 Kbits/sec   18             sender
[  5]   0.00-10.00  sec  15.7 KBytes  12.8 Kbits/sec                  receiver

dwmw2 replied:
I get about 800Mb/s sending from the client, 940Mb/s sending from the BPi.

dwmw2 replied:
Can you capture with tcpdump on the PC and share the cap file?

cioby23 replied:
I have compiled again today a new clean squashfs image for BPI-R2 target with your commits included and the situation is the same. I have saved a small tcpdump pcap file which captured the iperf test from the BananaPi R2 router to my local PC. I have uploaded the pcap file here https://pcap.honeynet.org.my/v1/submission/downloadpcap.php?id=12089

dwmw2 replied:
> have compiled again today a new clean squashfs image for BPI-R2 target with your commits included and the situation is the same.

Do you mean *without* my commits?

cioby23 replied:
No I mean I have applied all your commits to the OpenWRT master repository which I cloned it locally then compiled the image. Without your commits I think building the image won't work since the current OpenWRT master does not have u-boot and preloader packages added.

dwmw2 replied:
Right, but you can use the ones you have, and follow the existing (convoluted) install instructions on the OpenWrt hardware page for the R2.

I don't think my patches are the cause of this.

cioby23 replied:
I have tried that already to apply my patches to a newer version of the OpenWRT repo (If I remember correctly around 2 weeks ago and the issue is there). I agree also that your patches are not the cause for this problem. My working OpenWRT fork is from 2 months ago using kernel 5.4.31. During this time newer kernel versions were pushed are also some new patches were added here openwrt/target/linux/mediatek/patches-5.4/.

dwmw2 replied:
Should be relatively simple to bisect that and find which update broke it.

dwmw2 replied:
Let's discuss that elsewhere and close the discussion here.

dwmw2 replied:
@cioby23 Did you find the Ethernet problem? I think I'm seeing something similar on the U7623 now.

dwmw2 replied:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git/commit/?h=linux-5.4.y&id=1ff0732cf8055131050fa0c26c49429d6e725505 went into 5.4.34, which seems to be around the right time frame to suspect it. But it doesn't seem to be that.

@blogic any thoughts?

dwmw2 replied:
I dropped all of 0999-hnat.patch, 1000-eth-gdm-config-backport.patch and 1001-eth-mtk-disable-TBDDONE-support.patch too, which came in after the version you said was working. Those don't help either.


```
root@OpenWrt:/# iperf3 -c i7.infradead.org
Connecting to host i7.infradead.org, port 5201
[  5] local 2001:8b0:10b:1::c2a port 49498 connected to 2001:8b0:10b:1:21e:67ff:fecb:7a92 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   130 KBytes  1.06 Mbits/sec   26   1.39 KBytes       
[  5]   1.00-2.00   sec  0.00 Bytes  0.00 bits/sec    1   1.39 KBytes       
[  5]   2.00-3.00   sec  0.00 Bytes  0.00 bits/sec    1   1.39 KBytes       
[  5]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec    1   1.39 KBytes       
[  5]   4.00-5.00   sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
[  5]   5.00-6.00   sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
[  5]   6.00-7.00   sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
[  5]   7.00-8.00   sec  0.00 Bytes  0.00 bits/sec    1   1.39 KBytes       
[  5]   8.00-9.00   sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
[  5]   9.00-10.00  sec  0.00 Bytes  0.00 bits/sec    0   1.39 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   130 KBytes   106 Kbits/sec   30             sender
[  5]   0.00-10.00  sec  80.9 KBytes  66.2 Kbits/sec                  receiver

iperf Done.
root@OpenWrt:/# iperf3 -c i7.infradead.org -R
Connecting to host i7.infradead.org, port 5201
Reverse mode, remote host i7.infradead.org is sending
[  5] local 2001:8b0:10b:1::c2a port 49502 connected to 2001:8b0:10b:1:21e:67ff:fecb:7a92 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec   111 MBytes   927 Mbits/sec                  
[  5]   1.00-2.00   sec   110 MBytes   924 Mbits/sec                  
[  5]   2.00-3.00   sec   107 MBytes   899 Mbits/sec                  
[  5]   3.00-4.00   sec   107 MBytes   902 Mbits/sec                  
[  5]   4.00-5.00   sec   107 MBytes   899 Mbits/sec                  
[  5]   5.00-6.00   sec   107 MBytes   900 Mbits/sec                  
[  5]   6.00-7.00   sec   107 MBytes   900 Mbits/sec                  
[  5]   7.00-8.00   sec   107 MBytes   899 Mbits/sec                  
[  5]   8.00-9.00   sec   108 MBytes   902 Mbits/sec                  
[  5]   9.00-10.00  sec   108 MBytes   903 Mbits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  1.06 GBytes   907 Mbits/sec   54             sender
[  5]   0.00-10.00  sec  1.05 GBytes   905 Mbits/sec                  receiver

iperf Done.
```
```

dwmw2 replied:
Should be fixed in #3230 at least for Legacy IP. For IPv6 it's back to at least 2/3 of line rate, often more. In the 5.8.0-rc6 kernel I get 9xxMb/s in both directions for both IPv6 and Legacy IP with the patch from https://github.com/frank-w/BPI-R2-4.14/commit/ed15a1f84cfe2d731f20658b3c2dab3f7b3bc648

dwmw2 committed:
mediatek: mt7623: refresh 5.4 kernel config

Some options were not explicitly specified, causing the kernel build to
drop to interactive mode. Set the missing options.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
fstools: update to the latest version

d34ea8e Use autoclear for overlay loopback device

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

adrianschmutzler commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -1,12 +1,41 @@
 	mkf2fs e2fsprogs kmod-usb-ohci kmod-usb2 kmod-usb3 kmod-ata-ahci-mtk
+  UBOOT_TARGET := mt7623n_bpir2
+  IMAGES := img.gz
+  IMAGE/img.gz += banana-pi-sdcard | gzip | append-metadata
```
Is that "+=" intentional?

dwmw2 replied:
No. Killed; thanks.

dwmw2 committed:
mediatek: fix U-Boot pinctrl setup for mt7623 eMMC

The U-Boot pinctrl driver for mt7623 was incomplete and didn't handle the
settings required for eMMC to work.

Submitted upstream at
https://patchwork.ozlabs.org/project/uboot/list/?series=184529

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 commented:
@cioby23 If possible please could you show me the output of this command in U-Boot, when booted both from eMMC and from SD:
```
U-Boot> md 10005cc0 c
10005cc0: 00000511 00000511 00000511 00000511    ................
10005cd0: 00000411 00000411 00000411 00000411    ................
10005ce0: 00000011 00000011 00000011 00000011    ................
```
Also when U-Boot is started from SD, does `mmc dev 0` work for you? It shouldn't just silently drop back to a U-Boot prompt; it should look like this if it's working (which it doesn't for me right now):
```
U-Boot> mmc dev 0
switch to partitions #0, OK
mmc0(part 0) is current device
```

Then update to the latest build with the 'fix U-Boot pinctrl…' commit and repeat the same tests?

@kristrev this could do with being tested on the Unielec board too, please. Mine still hasn't arrived.

Thanks.

cioby23 commented:
@dwmw2 please find below the output of the **md** command when booted from SD card

```
U-Boot> md 10005cc0 c
10005cc0: 00000111 00000111 00000111 00000111    ................
10005cd0: 00000111 00000111 00000111 00000111    ................
10005ce0: 00000011 00000011 00000011 00000011    ................
```
For second question when booting from SD the **mmc** command output is displayed below
```
U-Boot> mmc dev 0
switch to partitions #0, OK
mmc0(part 0) is current device
```

dwmw2 committed:
mediatek: use U-Boot FAT environment support for Banana Pi R2

Instead of building in a default environment which loads our environment
from the FAT partition....  just ask U-Boot to do it.

Submitted upstream at
https://patchwork.ozlabs.org/project/uboot/list/?series=184688

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 commented:
OK, now that eMMC is working reliably when booted from SD I've also implemented a bootmenu option to install from the SD card to the internal eMMC.

ynezz commented:
tl;dr whats the status of this?

dwmw2 commented:
Waiting to be merged. I wanted to test that I haven't broken anything on the UniElec board, but that board has finally now arrived and appears to be hosed in master before I even touched it.

mans0n commented:
@dwmw2 How is it broken? Maybe it needs [this patch](https://github.com/openwrt/openwrt/pull/3066/commits/f14c4c55536e1dac13bfa484eb0c0e59760e6b79) to enable the second miniPCIe slot, and [this patch](https://github.com/openwrt/openwrt/pull/3066/commits/359b91d3b9edb23d583154c94857c0908ec76336) to activate the sysupgrade procedure.

dwmw2 commented:
It doesn't boot as described in https://bugs.openwrt.org/index.php?do=details&task_id=3217

Sysupgrade did kind of work from the vendor firmware to 19.07 but the MBR partition table needed to be fixed manually as described in https://forum.openwrt.org/t/cant-start-openwrt-19-07-2-on-unielec-u7623-02/61872/4

On the whole, I'm not going to lose sleep over the UniElec board. This PR is ready to merge for Banana Pi R2.

mans0n commented:
> It doesn't boot as described in https://bugs.openwrt.org/index.php?do=details&task_id=3217

So it definitely needs [this](https://github.com/openwrt/openwrt/pull/3066/commits/f14c4c55536e1dac13bfa484eb0c0e59760e6b79).

> Sysupgrade did kind of work from the vendor firmware to 19.07 but the MBR partition table needed to be fixed manually as described in https://forum.openwrt.org/t/cant-start-openwrt-19-07-2-on-unielec-u7623-02/61872/4

Isn't it already described in the device support commit message? 4def81f30f2e

> On the whole, I'm not going to lose sleep over the UniElec board. This PR is ready to merge for Banana Pi R2.

It might be, but you are also touching UniElec in this PR so it would be better to test it on a real device.

And I'd be glad if you add more description for commits that lack proper commit messages...

dwmw2 commented:
> So it definitely needs [this](https://github.com/openwrt/openwrt/pull/3066/commits/f14c4c55536e1dac13bfa484eb0c0e59760e6b79).

Not sure that would explain the failure mode. I think there's a pinctrl problem and even that doesn't explain a complete lockup. But I'll try it; thanks.

> > Sysupgrade did kind of work from the vendor firmware to 19.07 but the MBR partition table needed to be fixed manually as described in https://forum.openwrt.org/t/cant-start-openwrt-19-07-2-on-unielec-u7623-02/61872/4
> 
> Isn't it already described in the device support commit message? [4def81f](https://github.com/openwrt/openwrt/commit/4def81f30f2e89eeecc6ec06e3091f1dd6b3c3d9)

Not in a form that users could cope with. I'd like to make it automatic somehow. If we can build our own u-boot then we own the whole contents of the main eMMC device, since the preloader is on the boot partition anyway. And even if not we can carefully skip over the vendor one for now. I'm just really wary of messing with it since the board will be unrecoverable if I mess it up.

> > On the whole, I'm not going to lose sleep over the UniElec board. This PR is ready to merge for Banana Pi R2.
> 
> It might be, but you are also touching UniElec in this PR so it would be better to test it on a real device.

I'm not intentionally changing anything; only adding the BPI-R2 cases alongside it and some minor cosmetic tweaks with no (intentional) functional change.

> And I'd be glad if you add more description for commits that lack proper commit messages...

Anything in particular you feel is insufficiently described?

ynezz commented:
> Anything in particular you feel is insufficiently described?

Missing commit descriptions. I would need to add them myself, I don't push commits without commit descriptions.

dwmw2 commented:
Hm? They all have commit descriptions.

Some of those commit descriptions say everything that needs to be said in the first line without redundantly repeating themselves, but they all seem perfectly sufficient to me.

Is there any particular *information* you believe is missing? Any specific commit you looked at and thought, *"I don't know what's going on here..."*?

adrianschmutzler commented:
> Hm? They all have commit descriptions.

commit title != commit description

The first three commits I looked at only had title and SoB, so no "description" to that strict terms.

The description should give the reasons why the change needs to be done; e.g. for "mediatek: enable SATA for mt7623" you could just throw one or two sentences why SATA should be enabled for that platform (or maybe why it has not been enabled before and why you change that now).

I admit that there are better examples to explain that concept, though.

dwmw2 commented:
I agree, that isn't a stunning example. There is absolutely nothing relevant missing from that commit comment.

It enables SATA on the MT7623 SoC. I've been writing commit messages for a number of years now, and that is the important part.

I could add more words. I could explain the whole concept of a SoC, with all its standard associated set of peripherals. I could explain that SATA wasn't enabled before I enabled it, and I could explain that I enabled it because it's present on the SoC and because I think someone might want to use it.

But really, all that would be utterly pointless because it's implicit. I didn't enable SATA on a SoC that *doesn't* have it. I didn't enable it when I *didn't* want to use it. I wasn't *fixing* a previous attempt to enable SATA that hadn't quite worked for some reason.

I'm just enabling SATA on the MT7623.

If there are better examples **in this PR** where there is actual relevant information missing, please point them out. I'm happy to add missing information; just struggling with being asked for "more" when in my decades of writing commit messages my judgement is that I have nothing more to add without *detracting* from the succinctness of what has been said.

mans0n commented:
> I could add more words. I could explain the whole concept of a SoC, with all its standard associated set of peripherals. I could explain that SATA wasn't enabled before I enabled it, and I could explain that I enabled it because it's present on the SoC and because I think someone might want to use it.

Yes, that's what should be in the commit message. You can mention that the devices have a sata port that was not enabled before.

> But really, all that would be utterly pointless because it's implicit.

It's implicit for someone who has an experience on the mt7623 platform, but it might not be for someone else. For example, one might have difficulty understanding why the patch "mediatek: store MAC address in boot partition on Banana Pi R2" is needed without reading the UniElec support commit message. A good commit message should reduce such additional research.

Just one or two sentences would be fine. I admit that sometimes it feels quite bothering, but IMO it serves for good in the end.

dwmw2 commented:
> Yes, that's what should be in the commit message. You can mention that the devices have a sata port that was not enabled before.

But I did. That's what *"enable SATA for mt7623"* says.

Are you suggesting it should say something like, *"enable SATA (which exists, but wasn't enabled before) for mt7623"* ?

Is the bit within the parentheses really not *entirely* redundant?

For the MAC address, isn't that such a common thing that it doesn't need much explanation? But I concede I could add something along the lines of *"like the U7623 board, which also doesn't store the MAC address somewhere permanent and needs to remember the first randomly-generated one for ongoing consistency"*. It just seemed a little pointless.

mans0n commented:
> Are you suggesting it should say something like, "enable SATA (which exists, but wasn't enabled before) for mt7623"?

You can refer to 9e129c9d9548.

> But I concede I could add something along the lines of "like the U7623 board, which also doesn't store the MAC address somewhere permanent and needs to remember the first randomly-generated one for ongoing consistency". It just seemed a little pointless.

I don't think it's redundant, but feel free to disagree. I'm not a core dev here anyway.

bobafetthotmail commented:
> 
> Are you suggesting it should say something like, _"enable SATA (which exists, but wasn't enabled before) for mt7623"_ ?
> 

No that's not an answer to the "why" question, it just rephrases the same sentence.
That commit should say why you are doing it. For example "Added mt7623 device that uses Sata port, enable Sata support".
Or "these mt7623 devices have a Sata port, add sata support to their default package list"

> Is the bit within the parentheses really not _entirely_ redundant?
> 

no, as "enable sata" does not explain why you are doing it. Being a bit more veerbose helps reviewers. For example if you state that these devices have a Sata port they don't need to check themselves if the devices you are adding Sata support to actually have a port or not to decide if this commit is OK or not.

> For the MAC address, isn't that such a common thing that it doesn't need much explanation? But I concede I could add something along the lines of _"like the U7623 board, which also doesn't store the MAC address somewhere permanent and needs to remember the first randomly-generated one for ongoing consistency"_. It just seemed a little pointless.

In addition to the reasons above, git log (or commit history on Github web interface) is looked at and searchable by many types of users, not everyone has full understanding of OpenWrt build system and the history of the devices affected, and looking at commits can help other people learn what this code is doing.

That's why in the patch submission guidelines https://openwrt.org/submitting-patches it's required to write some form of commit description.

dwmw2 commented:
Honestly, if I'm enabling SATA on a device that doesn't actually have SATA, we have bigger problems than discussing pointless redundancy in commit messages.

And if I *believe* it has SATA but it really doesn't, that's still a concern. Me *stating* that it has SATA (in that case, wrongly) would still need to be caught be a reviewer.

And if I *know* it doesn't have SATA but I'm submitting this anyway, that's even weirder.

Saying "enable SATA (and yes, this SoC has SATA)" is just utterly pointless.

bobafetthotmail commented:
Honestly, you may disagree with the rules, but if you got the same feedback from both the guys that usually merge PRs on github you really should do that no matter how silly you think it is.

dwmw2 commented:
I have added more words. I believe them all to be fairly pointless and redundant, but it's easier than adding more words here.

Again, if there's anything actually *relevant* and *specific* that anyone was looking for, please feel free to ask.

But if it was just "do more pointless typing, regardless of the information being conveyed", I hope this suffices.

tmn505 commented:
Hi, have You considered writing U-Boot environment with this https://gitlab.denx.de/u-boot/u-boot/-/blob/master/doc/README.distro document in mind? It would help removing some of the logic in `mt7623n_bpir2-uEnv.txt`.  Assuming U-Boot enumerates always the SD card and eMMC with the same number.

dwmw2 commented:
> Hi, have You considered writing U-Boot environment with this https://gitlab.denx.de/u-boot/u-boot/-/blob/master/doc/README.distro document in mind?

Not really. We're building and shipping our own u-boot here, and we depend on certain features which I've deliberately built in; we can't do the "lowest common denominator" thing. The fun part of the u-boot script is the bit where it installs to the internal eMMC; the rest is in the noise.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

dwmw2 commented:
> So it definitely needs this.

@mans0n you're right, that patch does make it boot. It still has no Ethernet connectivity. Just an eth0 which receives no packets.

dwmw2 commented:
http://buildbot.openwrt.org/master/images/builders/mediatek%2Fmt7623/builds/448/steps/images/logs/stdio

```
./gen_banana_pi_img.sh: 81: ./gen_banana_pi_img.sh: Syntax error: "(" unexpected
```

Looks like there's a bashism in there after all. Will fix...
